test/simulator/main/org/apache/cassandra/simulator/paxos/PairOfSequencesAccordSimulation.java

```diff
@@ -281,21 +211,151 @@ protected String preInsertStmt()
     }
 
     @Override
-    Operation verifying(int operationId, IInvokableInstance instance, int primaryKey, HistoryChecker historyChecker)
+    boolean allowMultiplePartitions() { return true; }
+
+    @Override
+    BiFunction<SimulatedSystems, int[], Supplier<Action>> actionFactory()
     {
-        return new VerifyingOperation(operationId, instance, serialConsistency, primaryKey, historyChecker);
+        AtomicInteger id = new AtomicInteger(0);
+
+        return (simulated, primaryKeyIndex) -> {
+            int[] partitions = IntStream.of(primaryKeyIndex).map(i -> primaryKeys[i]).toArray();
+            return () -> accordAction(id.getAndIncrement(), simulated, partitions);
+        };
     }
 
-    @Override
-    Operation nonVerifying(int operationId, IInvokableInstance instance, int primaryKey, HistoryChecker historyChecker)
+    private static IIsolatedExecutor.SerializableCallable<SimpleQueryResult> query(int id, int[] partitions, int[] readOnly)
     {
-        return new NonVerifyingOperation(operationId, instance, serialConsistency, primaryKey, historyChecker);
+        return () -> execute(createAccordTxn(id, partitions, readOnly), "pk", "count", "seq");
     }
 
-    @Override
-    Operation modifying(int operationId, IInvokableInstance instance, int primaryKey, HistoryChecker historyChecker)
+    public class ReadWriteOperation extends Operation
     {
-        return new ModifyingOperation(operationId, instance, ANY, serialConsistency, primaryKey, historyChecker);
+        private final IntSet allPartitions;
+        private final IntSet readOnlySet;
```

Comment:
we should also have `writeOnly`?

Reply:
`allPartitions` contains all read/write partitions.  `readOnlySet` is the set that only did a read. Because we have access to the read accord did, all write operations have a read as well, so currently we do not get write only partitions.

I am not 100% sure how to do a write only txn as the validation logic tracks history; so if a txn doesn't load that history then its doing a blind write and could trigger a violation that wasn't a real violation.  My first thought would be that we periodically truncate the history, but do wonder if that would just make things more complex (if we have not observed all writes and did a truncate, may miss a violation)

Reply:
Well, the list append can be truly blind (and should be), so we’re testing additional functionality. If it isn’t today, that’s something we should address (it is for Paxos, and we shouldn’t regress).

We shouldn’t also witness any reads performed internally, and if we do that’s also a bug - and for now at least we should filter them if it isn’t trivial to fix properly.

I’m not sure what you mean about false violations but the verifier supports blind writes just fine?

Reply:
> Well, the list append can be truly blind (and should be), so we’re testing additional functionality. If it isn’t today, that’s something we should address (it is for Paxos, and we shouldn’t regress).

Makes sense, we can't do for `String` but *should* be able to do for `list`, so could leverage that.

> We shouldn’t also witness any reads performed internally, and if we do that’s also a bug 

Its an internal detail, normal CQL is unable to do this.  If you work at the `Txn` level you see the reads, if you work at the CQL level you only see what you asked for.

> I’m not sure what you mean about false violations but the verifier supports blind writes just fine?

its more if you are working with a type that requires read to update, then to avoid a read you need to ignore the history; simple example

```
create table foo (pk int PRIMARY KEY, seq text)

txn(INSERT INTO foo(pk, seq) VALUES (?, "lacking history...")
```

Reply:
Have you looked at how the Paxos one works? It uses `+=`, and can manage blind writes just fine for both registers. Obviously under the hood the implementation has to perform a read for the string based `+=`, so it is not a blind write in implementation. Saying that, it would probably be good to optionally flag `seq1` as maintained/unmaintained so we can separately track the correctness of truly blind writes, as with Accord they are treated differently.

Either way, we should definitely support UX blind writes, as these are different in implementation detail, even if only subtly.

Reply:
> Have you looked at how the Paxos one works? It uses +=, and can manage blind writes just fine for both registers. Obviously under the hood the implementation has to perform a read for the string based +=, so it is not a blind write in implementation. But at the CQL level it is a blind write, and at the implementation level these are treated slightly differently.


Yep, looked closely at it before doing this patch.  A blind write there still performs a read, so I took advantage of the fact we have access to the reads.  If we want to remove this direct access, then I can have `writeOnly` as the current writes are in fact blind...

If we are cool keeping this access, I did confirm `org.apache.cassandra.cql3.Lists.Appender#doAppend` doesn't do a read, but we need to make sure not flatten else it looks to just override the celll (aka does *not* append, it *sets*)... given that I can switch/add list to add blind writes.

> Saying that, it would probably also be good to randomise whether seq1 as maintained/unmaintained so we can separately track the correctness of truly blind writes, as with Accord they are treated differently.

I don't know what paxos does, can you explain?  Given `seq1 text` I assume this is doing a `set` rather than `append`?  So would need to extend the validators to handle this truncate of history?  How does paxos handle this differently?

Reply:
It does not do a read at the CQL layer, only internally. From the CQL perspective it is a blind write for Paxos.

We _shouldn't_ have access to the read in Accord if we perform a blind write either, as it shouldn't be returned to us since we didn't ask for it. We should be verifying this behaviour, even if it is only a modest divergence. Perhaps a small minority of writes should be run as blind at the CQL layer, but we should run at least some.

> I don't know what paxos does, can you explain? Given seq1 text I assume this is doing a set rather than append? So would need to extend the validators to handle this truncate of history? How does paxos handle this differently?

In this instance I am proposing additional functionality compared to the Paxos test, namely for some test runs we might want to randomly disable `seq1` entirely for the test duration, operating only on `seq2`, i.e. maintaining only seq2 and reading only seq2. This would permit us to run entirely on _true_ blind writes, to confirm they behave correctly, as they diverge more in implementation than with Paxos.

In the future, we might want to modify these tests not to rely on the reified history, by at least permitting history to be truncated. This is definitely possible with the strict serialisability verifier as it is designed, if we were to also request timestamps on read. But that's for another time.

Reply:
> We shouldn't have access to the read in Accord if we perform a blind write either, as it shouldn't be returned to us since we didn't ask for it. We should be verifying this behaviour, even if it is only a modest divergence.

I will update, I went this way as reading multiple rows is really annoying in the CQL layer atm...  If we had multi select statement support this feature would be nicer...

Reply:
Created https://issues.apache.org/jira/browse/CASSANDRA-18154 to support moving to CQL rather than always bypass...  This will allow the following options: read-only, write-only, read-write... Ill work on this change now.

Reply:
Moved to CQL which let me have : `writeOnly`, `readOnly`, `readWrite`